README.rst
###################
dorayaki (どら焼き)
###################


rebar deps::

    {dorayaki,
     ".*", {git, "git@github.com:shiguredo/dorayaki.git", {branch, "develop"}}},


dora_validate
    Simple validate library


CHANGES.rst
#####################
変更履歴
#####################

CHANGE
    下位互換のない変更
UPDATE
    下位互換がある変更
ADD
    下位互換がある追加
FIX
    バグ修正

develop
=======

1.0.0
=====

6 年越しの正式版リリース。


test/dora_datetime_tests.erl
-module(dora_datetime_tests).

-include_lib("eunit/include/eunit.hrl").

-include("dora_datetime.hrl").

-import(dora_datetime, [timestamp/2,
                        year/1, month/1, day/1,
                        hour/1, minute/1, second/1, micro_second/1, tz_offset/1,
                        iso8601/1, iso8601_no_micros/1, iso8601_no_millis/1, yyyymmdd/1,
                        iso8601_to_timestamp/1,
                        relativedelta/2,
                        posix_time_to_iso8601/1]).


timestamp_test() ->
    Timestamp = timestamp({1384, 213558, 1}, ?DORA_TZ_UTC),
    ?assertEqual(2013, year(Timestamp)),
    ?assertEqual(11, month(Timestamp)),
    ?assertEqual(11, day(Timestamp)),
    ?assertEqual(23, hour(Timestamp)),
    ?assertEqual(45, minute(Timestamp)),
    ?assertEqual(58, second(Timestamp)),
    ?assertEqual(1, micro_second(Timestamp)),
    ?assertEqual(0, tz_offset(Timestamp)),
    ok.


relativedelta_test() ->
    Timestamp = timestamp({1384, 213558, 1}, ?DORA_TZ_UTC),
    ?assertEqual(#dora_timestamp{year = 2013,
                                 month = 12,
                                 day = 1,
                                 hour = 23,
                                 minute = 45,
                                 second = 58,
                                 micro_second = 1,
                                 tz_offset = 0},
                 relativedelta(Timestamp, {days, 20})),
    ?assertEqual(#dora_timestamp{year = 2013,
                                 month = 11,
                                 day = 10,
                                 hour = 23,
                                 minute = 45,
                                 second = 58,
                                 micro_second = 1,
                                 tz_offset = 0},
                 relativedelta(Timestamp, {days, -1})),
    ?assertEqual(#dora_timestamp{year = 2013,
                                 month = 11,
                                 day = 11,
                                 hour = 22,
                                 minute = 45,
                                 second = 58,
                                 micro_second = 1,
                                 tz_offset = 0},
                 relativedelta(Timestamp, {hours, -1})),
    ?assertEqual(#dora_timestamp{year = 2013,
                                 month = 11,
                                 day = 12,
                                 hour = 0,
                                 minute = 45,
                                 second = 58,
                                 micro_second = 1,
                                 tz_offset = 0},
                 relativedelta(Timestamp, {hours, 1})),
    ok.


iso8601_test_() ->
    [
     ?_assertEqual(<<"2013-11-11T23:45:58.000000Z">>, iso8601(timestamp({1384, 213558, 0}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"2013-11-11T23:45:58.000001Z">>, iso8601(timestamp({1384, 213558, 1}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"2013-11-12T08:45:58.000001+09:00">>, iso8601(timestamp({1384, 213558, 1}, ?DORA_TZ_JST))),
     ?_assertEqual(<<"2013-11-11T18:45:58.000001-05:00">>, iso8601(timestamp({1384, 213558, 1}, ?DORA_TZ_ECT)))
    ].


iso8601_no_micros_test_() ->
    [
     ?_assertEqual(<<"2013-11-11T23:45:58.123Z">>, iso8601_no_micros(timestamp({1384, 213558, 123456}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"2013-11-11T23:45:58.123Z">>, iso8601_no_micros(timestamp({1384, 213558, 123456}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"2013-11-11T23:45:58.000Z">>, iso8601_no_micros(timestamp({1384, 213558, 1000000}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"2013-11-12T08:45:58.123+09:00">>, iso8601_no_micros(timestamp({1384, 213558, 123456}, ?DORA_TZ_JST))),
     ?_assertEqual(<<"2013-11-11T18:45:58.123-05:00">>, iso8601_no_micros(timestamp({1384, 213558, 123456}, ?DORA_TZ_ECT)))
    ].

iso8601_no_millis_test_() ->
    [
     ?_assertEqual(<<"2013-11-11T23:45:58Z">>, iso8601_no_millis(timestamp({1384, 213558, 123456}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"2013-11-11T23:45:58Z">>, iso8601_no_millis(timestamp({1384, 213558, 123456}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"2013-11-12T08:45:58+09:00">>, iso8601_no_millis(timestamp({1384, 213558, 123456}, ?DORA_TZ_JST))),
     ?_assertEqual(<<"2013-11-11T18:45:58-05:00">>, iso8601_no_millis(timestamp({1384, 213558, 123456}, ?DORA_TZ_ECT)))
    ].

iso8601_to_timestamp_test_() ->
    [
     ?_assertEqual({error, {unsupported_format, <<"-">>}}, iso8601_to_timestamp(<<"-">>)),
     ?_assertEqual({error, {unsupported_date_format, <<"201-11-11">>}}, iso8601_to_timestamp(<<"201-11-11T23:45:58.000000">>)),
     ?_assertEqual({error, {unsupported_time_format, <<"23:45:5.000000">>}}, iso8601_to_timestamp(<<"2013-11-11T23:45:5.000000">>)),

     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.000000">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.000001">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.000000Z">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.000001Z">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1}, ?DORA_TZ_JST)}, iso8601_to_timestamp(<<"2013-11-12T08:45:58.000001+09:00">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1}, ?DORA_TZ_JST)}, iso8601_to_timestamp(<<"2013-11-12T08:45:58.000001+0900">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1}, ?DORA_TZ_ECT)}, iso8601_to_timestamp(<<"2013-11-11T18:45:58.000001-05:00">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1}, ?DORA_TZ_ECT)}, iso8601_to_timestamp(<<"2013-11-11T18:45:58.000001-0500">>)),

     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.000">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1000}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.001">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.000Z">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1000}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58.001Z">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1000}, ?DORA_TZ_JST)}, iso8601_to_timestamp(<<"2013-11-12T08:45:58.001+09:00">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1000}, ?DORA_TZ_JST)}, iso8601_to_timestamp(<<"2013-11-12T08:45:58.001+0900">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1000}, ?DORA_TZ_ECT)}, iso8601_to_timestamp(<<"2013-11-11T18:45:58.001-05:00">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 1000}, ?DORA_TZ_ECT)}, iso8601_to_timestamp(<<"2013-11-11T18:45:58.001-0500">>)),

     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58Z">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_UTC)}, iso8601_to_timestamp(<<"2013-11-11T23:45:58Z">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_JST)}, iso8601_to_timestamp(<<"2013-11-12T08:45:58+09:00">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_JST)}, iso8601_to_timestamp(<<"2013-11-12T08:45:58+0900">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_ECT)}, iso8601_to_timestamp(<<"2013-11-11T18:45:58-05:00">>)),
     ?_assertEqual({ok, timestamp({1384, 213558, 0}, ?DORA_TZ_ECT)}, iso8601_to_timestamp(<<"2013-11-11T18:45:58-0500">>))
    ].

yyyymmdd_test_() ->
    [
     ?_assertEqual(<<"20131111">>, yyyymmdd(timestamp({1384, 213558, 1}, ?DORA_TZ_UTC))),
     ?_assertEqual(<<"20131112">>, yyyymmdd(timestamp({1384, 213558, 1}, ?DORA_TZ_JST)))
    ].



posix_time_to_iso8601_test_() ->
    [
     ?_assertEqual(<<"1998-12-31T23:59:59Z">>, posix_time_to_iso8601(915148799)),
     ?_assertEqual(<<"2038-01-19T03:14:07Z">>, posix_time_to_iso8601(2147483647))
    ].


test/dora_validate_tests.erl
-module(dora_validate_tests).

-import(dora_validate, [validate/2]).

-include_lib("eunit/include/eunit.hrl").


validate_type_ipv4_address_and_port_number_test() ->
    application:set_env(dorayaki, node, "127.0.0.1:5000"),
    ?assertEqual(ok,
                 validate(dorayaki, [{node, ipv4_address_and_port_number, required}])),

    application:set_env(dorayaki, node, "dorayaki"),
    ?assertEqual({error, {badarg, node, ipv4_address_and_port_number, "dorayaki"}},
                 validate(dorayaki, [{node, ipv4_address_and_port_number, required}])),

    ?assertEqual(ok,
                 application:unset_env(dorayaki, ip_address)),
    ok.


validate_type_list_ipv4_address_and_port_number_test() ->
    application:set_env(dorayaki, node_list, ["192.0.2.1:5000", "192.0.2.2:5000"]),
    ?assertEqual(ok,
                 validate(dorayaki, [{node_list, list_ipv4_address_and_port_number, required}])),

    application:set_env(dorayaki, node_list, ["dorayaki", "192.0.2.1:5000"]),
    ?assertEqual({error, {badarg, node_list, list_ipv4_address_and_port_number, ["dorayaki", "192.0.2.1:5000"]}},
                 validate(dorayaki, [{node_list, list_ipv4_address_and_port_number, required}])),

    ?assertEqual(ok,
                 application:unset_env(dorayaki, ip_address)),
    ok.


validate_type_ipv4_address_test() ->
    %% optional
    ?assertEqual(ok,
                 validate(dorayaki, [{ip_address, ipv4_address, optional}])),
    ?assertEqual(undefined,
                 application:get_env(dorayaki, ip_address)),

    %% optional default
    ?assertEqual(ok,
                 validate(dorayaki, [{ip_address, ipv4_address, optional, {127,0,0,1}}])),
    ?assertEqual({ok, {127,0,0,1}},
                 application:get_env(dorayaki, ip_address)),

    application:set_env(dorayaki, ip_address, "127.0.0.1"),
    ?assertEqual(ok,
                 validate(dorayaki, [{ip_address, ipv4_address, required}])),

    application:set_env(dorayaki, ip_address, "abc"),
    ?assertEqual({error, {badarg, ip_address, ipv4_address, "abc"}},
                 validate(dorayaki, [{ip_address, ipv4_address, required}])),

    application:set_env(dorayaki, ip_address, "::1"),
    ?assertEqual({error, {badarg, ip_address, ipv4_address, "::1"}},
                 validate(dorayaki, [{ip_address, ipv4_address, required}])),

    ?assertEqual(ok,
                 application:unset_env(dorayaki, ip_address)),
    ok.


validate_type_list_ipv4_address_test() ->
    %% optional
    ?assertEqual(ok,
                 validate(dorayaki, [{listen_ipv4_address, {listen_ipv4_address, 1}, optional}])),
    ?assertEqual(undefined,
                 application:get_env(dorayaki, listen_ipv4_address)),

    %% optional default
    ?assertEqual(ok,
                 validate(dorayaki, [{listen_ipv4_address, {list_ipv4_address, 1}, optional, [{127,0,0,1}]}])),
    ?assertEqual({ok, [{127,0,0,1}]},
                 application:get_env(dorayaki, listen_ipv4_address)),

    application:set_env(dorayaki, listen_ipv4_address, ["127.0.0.1", "192.168.1.10"]),
    ?assertEqual(ok,
                 validate(dorayaki, [{listen_ipv4_address, {list_ipv4_address, 1}, required}])),

    application:set_env(dorayaki, listen_ipv4_address, ["127.0.0.1", "192.168.1.10"]),
    ?assertEqual({error, {badarg, listen_ipv4_address, {list_ipv4_address, 3}, ["127.0.0.1", "192.168.1.10"]}},
                 validate(dorayaki, [{listen_ipv4_address, {list_ipv4_address, 3}, required}])),

    application:set_env(dorayaki, listen_ipv4_address, ["abc"]),
    ?assertEqual({error, {badarg, listen_ipv4_address, {list_ipv4_address, 1}, ["abc"]}},
                 validate(dorayaki, [{listen_ipv4_address, {list_ipv4_address, 1}, required}])),

    application:set_env(dorayaki, listen_ipv4_address, ["::1"]),
    ?assertEqual({error, {badarg, listen_ipv4_address, {list_ipv4_address, 1}, ["::1"]}},
                 validate(dorayaki, [{listen_ipv4_address, {list_ipv4_address, 1}, required}])),

    ?assertEqual(ok,
                 application:unset_env(dorayaki, listen_ipv4_address)),
    ok.


validate_type_ipv6_address_test() ->
    %% optional
    ?assertEqual(ok,
                 validate(dorayaki, [{ipv6_address, ipv6_address, optional}])),
    ?assertEqual(undefined,
                 application:get_env(dorayaki, ipv6_address)),

    %% optional default
    ?assertEqual(ok,
                 validate(dorayaki, [{ipv6_address, ipv6_address, optional, {0,0,0,0,0,0,0,1}}])),
    ?assertEqual({ok, {0,0,0,0,0,0,0,1}},
                 application:get_env(dorayaki, ipv6_address)),

    application:set_env(dorayaki, ipv6_address, "2001:e42:102:1527:160:16:103:161"),
    ?assertEqual(ok,
                 validate(dorayaki, [{ipv6_address, ipv6_address, required}])),

    ?assertEqual(ok,
                 application:unset_env(dorayaki, ip_address)),
    ok.


validate_type_list_ipv6_address_test() ->
    %% optional
    ?assertEqual(ok,
                 validate(dorayaki, [{listen_ipv6_address, {list_ipv6_address, 1}, optional}])),
    ?assertEqual(undefined,
                 application:get_env(dorayaki, listen_ipv6_address)),

    %% optional default
    ?assertEqual(ok,
                 validate(dorayaki, [{listen_ipv6_address, {list_ipv6_address, 1}, optional, [{0,0,0,0,0,0,0,1}]}])),
    ?assertEqual({ok, [{0,0,0,0,0,0,0,1}]},
                 application:get_env(dorayaki, listen_ipv6_address)),

    application:set_env(dorayaki, listen_ipv6_address, ["2001:e42:102:1527:160:16:103:161", "::1"]),
    ?assertEqual(ok,
                 validate(dorayaki, [{listen_ipv6_address, {list_ipv6_address, 0}, required}])),

    application:set_env(dorayaki, listen_ipv6_address, ["2001:e42:102:1527:160:16:103:161", "::1"]),
    ?assertEqual({error, {badarg, listen_ipv6_address, {list_ipv6_address, 3},
                          ["2001:e42:102:1527:160:16:103:161", "::1"]}},
                 validate(dorayaki, [{listen_ipv6_address, {list_ipv6_address, 3}, required}])),

    ?assertEqual(ok,
                 application:unset_env(dorayaki, ip_address)),
    ok.



validate_type_port_number_test() ->
    %% optional
    ?assertEqual(ok,
        validate(dorayaki, [{port_number, port_number, optional}])),
    ?assertEqual(undefined,
        application:get_env(dorayaki, port_number)),

    %% optional default
    ?assertEqual(ok,
        validate(dorayaki, [{port_number, port_number, optional, 1}])),
    ?assertEqual({ok, 1},
        application:get_env(dorayaki, port_number)),

    application:set_env(dorayaki, port_number, 1),
    ?assertEqual(ok,
        validate(dorayaki, [{port_number, port_number, required}])),

    application:set_env(dorayaki, port_number, 65535),
    ?assertEqual(ok,
        validate(dorayaki, [{port_number, port_number, required}])),

    %% bad range
    application:set_env(dorayaki, port_number, -1),
    ?assertEqual({error, {badarg, port_number, port_number, -1}},
        validate(dorayaki, [{port_number, port_number, required}])),

    application:set_env(dorayaki, port_number, 65536),
    ?assertEqual({error, {badarg, port_number, port_number, 65536}},
        validate(dorayaki, [{port_number, port_number, required}])),

    %% bad type
    application:set_env(dorayaki, port_number, "abc"),
    ?assertEqual({error, {badarg, port_number, port_number, "abc"}},
        validate(dorayaki, [{port_number, port_number, required}])),

    ?assertEqual(ok,
        application:unset_env(dorayaki, port_number)),
    ok.


validate_type_http_uri_test() ->
    %% optional
    ?assertEqual(ok,
        validate(dorayaki, [{webhook_url, http_uri, optional}])),
    ?assertEqual(undefined,
        application:get_env(dorayaki, webhook_url)),

    %% optional default
    ?assertEqual(ok,
        validate(dorayaki, [{webhook_url, http_uri, optional, []}])),
    ?assertEqual({ok, []},
        application:get_env(dorayaki, webhook_url)),

    %% https://
    application:set_env(dorayaki, webhook_url, "http://example.com"),
    ?assertEqual(ok,
        validate(dorayaki, [{webhook_url, http_uri, required}])),

    %% http://
    application:set_env(dorayaki, webhoo_url, "https//google.com"),
    ?assertEqual(ok,
        validate(dorayaki, [{webhook_url, http_uri, required}])),

    %% bad uri
    application:set_env(dorayaki, webhook_url, "ntp://example.com"),
    ?assertEqual({error, {badarg, webhook_url, http_uri, "ntp://example.com"}},
        validate(dorayaki, [{webhook_url, http_uri, required}])),

    %% bad type
    application:set_env(dorayaki, webhook_url, 1),
    ?assertEqual({error, {unknown_type, webhook_url, http_url, 1}},
                 %% http_ur*l*
        validate(dorayaki, [{webhook_url, http_url, required}])),

    ?assertEqual(ok,
        application:unset_env(dorayaki, webhook_url)),
    ok.


validate_type_list_list_http_uri_test() ->
    %% optional
    ?assertEqual(ok,
        validate(dorayaki, [{server_urls, list_http_uri, optional}])),
    ?assertEqual(undefined,
        application:get_env(dorayaki, server_urls)),

    %% https://
    application:set_env(dorayaki, server_urls, ["http://example.com"]),
    ?assertEqual(ok,
        validate(dorayaki, [{server_urls, list_http_uri, required}])),

    %% http://
    application:set_env(dorayaki, webhoo_url, ["https//google.com"]),
    ?assertEqual(ok,
        validate(dorayaki, [{server_urls, list_http_uri, required}])),

    %% bad uri
    application:set_env(dorayaki, server_urls, ["ntp://example.com"]),
    ?assertEqual({error, {badarg, server_urls, list_http_uri, ["ntp://example.com"]}},
        validate(dorayaki, [{server_urls, list_http_uri, required}])),

    %% bad type
    application:set_env(dorayaki, server_urls, 1),
    ?assertEqual({error, {unknown_type, server_urls, http_url, 1}},
                 %% http_ur*l*
        validate(dorayaki, [{server_urls, http_url, required}])),

    ?assertEqual(ok,
        application:unset_env(dorayaki, server_urls)),
    ok.





src/dora_datetime.erl
-module(dora_datetime).

-export([timestamp/0, timestamp/1, timestamp/2]).
-export([year/1, month/1, day/1,
         hour/1, minute/1, second/1, micro_second/1, tz_offset/1]).
-export([timestamp_to_datetime/1, timestamp_to_gregorian_seconds/1,
         datetime_to_timestamp/1]).
-export([iso8601/1, iso8601_no_micros/1, iso8601_no_millis/1, yyyymmdd/1,
         iso8601_to_timestamp/1]).
-export([relativedelta/2]).

-export([posix_time_to_iso8601/1]).

-export_type([month/0, day/0]).

-include("dora_datetime.hrl").

-type month() :: 1..12.
-type day() :: 1..31.

-spec year(#dora_timestamp{}) -> non_neg_integer().
year(#dora_timestamp{year = Year}) ->
    Year.


-spec month(#dora_timestamp{}) -> month().
month(#dora_timestamp{month = Month}) ->
    Month.


-spec day(#dora_timestamp{}) -> day().
day(#dora_timestamp{day = Day}) ->
    Day.


-spec hour(#dora_timestamp{}) -> 0..23.
hour(#dora_timestamp{hour = Hour}) ->
    Hour.


-spec minute(#dora_timestamp{}) -> 0..59.
minute(#dora_timestamp{minute = Minute}) ->
    Minute.


second(#dora_timestamp{second = Second}) ->
    Second.


micro_second(#dora_timestamp{micro_second = MicroSecond}) ->
    MicroSecond.


-spec tz_offset(#dora_timestamp{}) -> non_neg_integer().
tz_offset(#dora_timestamp{tz_offset = TzOffset}) ->
    TzOffset.


-spec relativedelta(#dora_timestamp{}, {atom(), integer()}) -> #dora_timestamp{}.
relativedelta(_Timestamp, {years, _Value}) ->
    error(not_implemented);
relativedelta(_Timestamp, {months, _Value}) ->
    error(not_implemented);
relativedelta(Timestamp, {days, Value}) ->
    GregorianSeconds = timestamp_to_gregorian_seconds(Timestamp),
    GregorianSeconds1 = GregorianSeconds + (24 * 60 * 60) * Value,
    {{Year, Month, Day}, {Hour, Minute, Second}} = calendar:gregorian_seconds_to_datetime(GregorianSeconds1),
    Timestamp#dora_timestamp{year = Year, month = Month, day = Day,
                             hour = Hour, minute = Minute, second = Second};
relativedelta(Timestamp, {hours, Value}) ->
    GregorianSeconds = timestamp_to_gregorian_seconds(Timestamp),
    GregorianSeconds1 = GregorianSeconds + (60 * 60) * Value,
    {{Year, Month, Day}, {Hour, Minute, Second}} = calendar:gregorian_seconds_to_datetime(GregorianSeconds1),
    Timestamp#dora_timestamp{year = Year, month = Month, day = Day,
                             hour = Hour, minute = Minute, second = Second};
relativedelta(_Timestamp, {minites, _Value}) ->
    error(not_implemented);
relativedelta(_Timestamp, {seconds, _Value}) ->
    error(not_implemented);
relativedelta(_Timestamp, {microseconds, _Value}) ->
    error(not_implemented).




-spec time_zone(atom()) -> {integer(), binary()}.
time_zone(?DORA_TZ_UTC) ->
    {0, <<"Z">>};
time_zone(?DORA_TZ_JST) ->
    {9, <<"+09:00">>};
time_zone(?DORA_TZ_ECT) ->
    {-5, <<"-05:00">>};
time_zone(Tz) ->
    error({not_implemented, Tz}).


-spec timestamp() -> #dora_timestamp{}.
timestamp() ->
    timestamp(?DORA_TZ_UTC).

-spec timestamp(atom()) -> #dora_timestamp{}.
timestamp(Tz) ->
    Timestamp = erlang:timestamp(),
    timestamp(Timestamp, Tz).

-spec timestamp(erlang:timestamp(), atom()) -> #dora_timestamp{}.
timestamp({_MegaSecs, _Secs, MicroSecs} = Timestamp, Tz) ->
    Datetime = calendar:now_to_universal_time(Timestamp),
    {Offset, Designator} = time_zone(Tz),
    Seconds = calendar:datetime_to_gregorian_seconds(Datetime) + (60 * 60 * Offset),
    {{Year, Month, Day}, {Hour, Minute, Second}} = calendar:gregorian_seconds_to_datetime(Seconds),
    #dora_timestamp{year = Year,
                    month = Month,
                    day = Day,
                    hour = Hour,
                    minute = Minute,
                    second = Second,
                    micro_second = MicroSecs,
                    tz_offset = Offset,
                    tz_designator = Designator}.


timestamp_to_datetime(#dora_timestamp{year = Year,
                                      month = Month,
                                      day = Day,
                                      hour = Hour,
                                      minute = Minute,
                                      second = Second}) ->
    {{Year, Month, Day}, {Hour, Minute, Second}}.


datetime_to_timestamp({{Year, Month, Day}, {Hour, Minute, Second}}) ->
    #dora_timestamp{year = Year,
                    month = Month,
                    day = Day,
                    hour = Hour,
                    minute = Minute,
                    second = Second,
                    micro_second = 0}.


timestamp_to_gregorian_seconds(Timestamp) when is_record(Timestamp, dora_timestamp) ->
    Datetime = timestamp_to_datetime(Timestamp),
    calendar:datetime_to_gregorian_seconds(Datetime).


-spec iso8601(#dora_timestamp{}) -> binary().
iso8601(#dora_timestamp{year = Year, month = Month, day = Day,
                        hour = Hour, minute = Minute, second = Second, micro_second = MicroSecs,
                        tz_designator = TzDesignator}) ->
    list_to_binary(io_lib:format("~4.10.0b-~2.10.0b-~2.10.0bT~2.10.0b:~2.10.0b:~2.10.0b.~6.10.0b~s",
                                 [Year, Month, Day, Hour, Minute, Second, MicroSecs, TzDesignator])).


iso8601_no_micros(#dora_timestamp{
                     year = Year, month = Month, day = Day,
                     hour = Hour, minute = Minute, second = Second,
                     micro_second = MicroSecs,
                     tz_designator = TzDesignator
                    }) ->
    Milli = erlang:round(MicroSecs / 1000) rem 1000,
    list_to_binary(io_lib:format("~4.10.0b-~2.10.0b-~2.10.0bT~2.10.0b:~2.10.0b:~2.10.0b.~3.10.0b~s",
                                 [Year, Month, Day, Hour, Minute, Second, Milli, TzDesignator])).



iso8601_no_millis(#dora_timestamp{year = Year, month = Month, day = Day,
                                  hour = Hour, minute = Minute, second = Second,
                                  tz_designator = TzDesignator}) ->
    list_to_binary(io_lib:format("~4.10.0b-~2.10.0b-~2.10.0bT~2.10.0b:~2.10.0b:~2.10.0b~s",
                                 [Year, Month, Day, Hour, Minute, Second, TzDesignator])).


-spec iso8601_to_timestamp(binary()) -> {ok, #dora_timestamp{}} | {error, unsupported_format}.
iso8601_to_timestamp(Binary) ->
    case binary:split(Binary, <<"T">>) of
        [Date, TimeAndTz] ->
            DateRegExp = <<"^(\\d{4})\-*(\\d{2})\-*(\\d{2})$">>,
            case re:run(Date, DateRegExp, [{capture, all_but_first, binary}]) of
                {match, [YearBin, MonthBin, DayBin]} ->
                    TimeRegExp = <<"^(\\d{2})\:*(\\d{2})\:*(\\d{2})([\.\,]\\d{3,6})?(\.*)$">>,
                    case re:run(TimeAndTz, TimeRegExp, [{capture, all_but_first, binary}]) of
                        {match, [HourBin, MinuteBin, SecondBin, <<>>, Tz]} ->
                            {ok, iso8601_to_timestamp(YearBin, MonthBin, DayBin,
                                                      HourBin, MinuteBin, SecondBin, <<"0">>, Tz)};
                        {match, [HourBin, MinuteBin, SecondBin, <<_:1/binary, MilliSecondBin:3/binary>>, Tz]} ->
                            {ok, iso8601_to_timestamp(YearBin, MonthBin, DayBin,
                                                      HourBin, MinuteBin, SecondBin,
                                                      <<MilliSecondBin/binary, "000">>, Tz)};
                        {match, [HourBin, MinuteBin, SecondBin, <<_:1/binary, MicroSecondBin:4/binary>>, Tz]} ->
                            {ok, iso8601_to_timestamp(YearBin, MonthBin, DayBin,
                                                      HourBin, MinuteBin, SecondBin, MicroSecondBin, Tz)};
                        {match, [HourBin, MinuteBin, SecondBin, <<_:1/binary, MicroSecondBin:5/binary>>, Tz]} ->
                            {ok, iso8601_to_timestamp(YearBin, MonthBin, DayBin,
                                                      HourBin, MinuteBin, SecondBin, MicroSecondBin, Tz)};
                        {match, [HourBin, MinuteBin, SecondBin, <<_:1/binary, MicroSecondBin:6/binary>>, Tz]} ->
                            {ok, iso8601_to_timestamp(YearBin, MonthBin, DayBin,
                                                      HourBin, MinuteBin, SecondBin, MicroSecondBin, Tz)};
                        _ ->
                            {error, {unsupported_time_format, TimeAndTz}}
                    end;
                _ ->
                    {error, {unsupported_date_format, Date}}
            end;
        _ ->
            {error, {unsupported_format, Binary}}
    end.


iso8601_to_timestamp(YearBin, MonthBin, DayBin, HourBin, MinuteBin, SecondBin, MicroSecondBin, Tz) ->
    Timestamp = #dora_timestamp{year = binary_to_integer(YearBin),
                                month = binary_to_integer(MonthBin),
                                day = binary_to_integer(DayBin),
                                hour = binary_to_integer(HourBin),
                                minute = binary_to_integer(MinuteBin),
                                second = binary_to_integer(SecondBin),
                                micro_second = binary_to_integer(MicroSecondBin)},
    case Tz of
        <<>> ->
            Timestamp;
        <<"Z">> ->
            Timestamp;
        <<"+", Hour:2/binary, _Rest/binary>> ->
            %% TODO(shimazaki): Rest が "00" か ":00" であることを確認する
            Timestamp#dora_timestamp{tz_offset = binary_to_integer(Hour),
                                     tz_designator = <<"+", Hour/binary, ":00">>};
        <<"-", Hour:2/binary, _Rest/binary>> ->
            %% TODO(shimazaki): Rest が "00" か ":00" であることを確認する
            Timestamp#dora_timestamp{tz_offset = -1 * binary_to_integer(Hour),
                                     tz_designator = <<"-", Hour/binary, ":00">>}
    end.


-spec yyyymmdd(calendar:date() | #dora_timestamp{}) -> binary().
yyyymmdd(#dora_timestamp{year = Year, month = Month, day = Day}) ->
    list_to_binary(io_lib:format("~4.10.0b~2.10.0b~2.10.0b", [Year, Month, Day])).



posix_time_to_iso8601(PosixTime) ->
    GregorianSeconds = calendar:datetime_to_gregorian_seconds({{1970,1,1}, {0,0,0}}) + PosixTime,
    Datetime = calendar:gregorian_seconds_to_datetime(GregorianSeconds),
    Timestamp = datetime_to_timestamp(Datetime),
    iso8601_no_millis(Timestamp).



